2370 Fixes recap email parse error

albertisfu opened the pull request:
This PR solves [#2370](https://github.com/freelawproject/courtlistener/issues/2370) 

- This error was raised when a recap.email notification containing attachments was sent two or more times for different recap.email users. 

  The problem was that the main recap document used to query the attachment page and merge attachments was obtained from `rds_created`.  But when a recap.email notification is sent twice (for different users) the second and the following notifications failed since there were no new `rds_created`.  To solve it now instead of using `rds_created` the main recap document is obtained from the related docket entry `des_returned` selecting the main recap document as the earliest related RD.

- Doing more related testing but using a multi-docket NEF notification with attachments I realized a similar error that might happen as a corner case:

  For multi-docket NEF we get the attachment report only for one docket entry and copy it to the other docket entries to avoid querying PACER multiple times since all other docket entries and their attachments are the same. 
  
  But if a multi-docket NEF with attachments is sent two or more times, an error when adding attachments is raised due to the RECAPDocument DB constraint for attachments.
  
  We download the attachment report based on the main `RECAPDocument` of the first docket entry. 
  
  Let's say the main RD with document number: 26
  
  The attachment report is the same for all docket entries, the only difference is the document number of every docket entry mentioned in the NEF. 
  
  So we use the same attachment report to merge attachments into each docket entry and then only update the corresponding document number to each docket entry.
  
  The problem occurred when a multi-docket NEF with attachments is received more than one time, due to each docket entry has already attachments. Since the attachment report uses the document number of the first docket entry (26), it doesn't match to current attachments in the docket entry so it's considered a new attachment to create.  Then when trying to update its corresponding document number an integrity error is raised since it already exists an attachment with the same document number.
  
  So in order to solve it, before getting attachments we check if it exists an attachment processing queue for that docket entry created by recap.email in the last hour (a cut-off time to consider is the same recap.email notification). If so, it means that the notification has already been sent before by another recap.email user, so we avoid to processes attachments again.

mlissner commented:
Good finds. The code looks fine, but the method is way too long. Can you think about ways to factor things out into helper functions?

> I think is not possible to receive a different recap.email notification for the same docket entry, I mean a notification that updates the previous one, like adding new attachments to the existing ones?

I could see updates for a particular docket entry coming in. For example, if a clerk makes a mistake, maybe they'll fix the mistake and send another notification. That could happen within a few minutes or maybe days later?

albertisfu commented:
Thanks for your comments, I've applied the suggested changes, moved some code out into a couple of helper functions to make the main method shorter and more readable.

About this:
> I could see updates for a particular docket entry coming in. For example, if a clerk makes a mistake, maybe they'll fix the mistake and send another notification. That could happen within a few minutes or maybe days later?

Right, in that case, I think it's better to remove the condition that checked if we have already processed attachments for this notification to avoid processing them again. Because if there's an additional notification due to a mistake from the court that adds or modifies the data for attachments now we'll be able to process them again. Now to solve the problem instead of overwriting the document number for attachments after adding them. I modified the `process_recap_attachment` method to accept as an optional parameter the document number, so we can assign the corresponding document number for each document when merging them and avoid duplicates.

**I added also in this PR the fix for: [2353](https://github.com/freelawproject/courtlistener/issues/2353)**

The problem is when receiving emails that are not valid notifications like:

![Screen Shot 2022-11-29 at 13 49 41](https://user-images.githubusercontent.com/486004/204633609-c4f28783-dc28-45aa-b230-33cbd5dafb2c.png)
![Screen Shot 2022-11-29 at 13 48 55](https://user-images.githubusercontent.com/486004/204633401-b9c98ae0-562d-4ff8-b757-d204dc17bfb6.png)

The solution was to add an additional rule to data validation that verifies the list of dockets contains at least one docket:
`len(data["dockets"]) == 0`

Let me know what you think.

mlissner commented on cl/recap/tasks.py:
```diff
@@ -1838,6 +1844,99 @@ class DocketUpdatedData:
+            user_pk,
+            att_report_text,
+        )
+        # We only query and parse the Attachment page for one document in PACER
+        # The only field that needs to be overwritten is the document number
+        # since it changes from case to case. Assigns it from its main document
```
Can you clarify that this to explain how it applies only to multi-docket notifications?

mlissner reviewed (changes requested):
This refactor is a lot nicer. I just think we need a slightly better comment, and we should be good.

albertisfu commented:
Yes of course, I've added more details to the comment about how why is needed to assign the document number when merging attachments for multi-docket nefs.
